shimmy/utils/dm_env_utils.py

```diff
@@ -7,40 +7,47 @@
 import dm_env
 import numpy as np
-from dm_env.specs import Array, BoundedArray, DiscreteArray
+import tree
 from gymnasium import spaces
 
 
-def dm_spec2gym_space(spec) -> spaces.Space[Any]:
-    """Converts a dm_env spec to a gymnasium space."""
-    if isinstance(spec, (OrderedDict, dict)):
-        return spaces.Dict(
-            {key: dm_spec2gym_space(value) for key, value in copy.copy(spec).items()}
-        )
-    # not possible to use isinstance due to inheritance
-    elif type(spec) is BoundedArray:
+def dm_spec2gym_space(spec: tree.Structure[dm_env.specs.Array]) -> spaces.Space[Any]:
+    """Converts a dm_env nested structure of specs to a Gymnasium Space.
+
+    BoundedArray is converted to Box Gymnasium spaces. DiscreteArray is converted to
+    Discrete Gymnasium spaces. Using Tuple and Dict spaces recursively as needed.
+
+    Args:
+        spec: The nested structure of specs
+
+    Returns:
+        The Gymnasium space corresponding to the given spec.
+    """
+    if isinstance(spec, dm_env.specs.DiscreteArray):
+        return spaces.Discrete(spec.num_values)
+    elif isinstance(spec, dm_env.specs.BoundedArray):
         low = np.broadcast_to(spec.minimum, spec.shape)
         high = np.broadcast_to(spec.maximum, spec.shape)
-        return spaces.Box(low=low, high=high, shape=spec.shape, dtype=spec.dtype)
-    elif type(spec) is Array:
-        if np.issubdtype(spec.dtype, np.integer):
-            low = np.iinfo(spec.dtype).min
-            high = np.iinfo(spec.dtype).max
-        elif np.issubdtype(spec.dtype, np.inexact):
-            low = float("-inf")
-            high = float("inf")
+        return spaces.Box(low, high, spec.shape, spec.dtype)
+    elif isinstance(spec, dm_env.specs.Array):
+        if np.issubdtype(spec.dtype, np.inexact):
+            return spaces.Box(-np.inf, np.inf, spec.shape, spec.dtype)
+        elif np.issubdtype(spec.dtype, np.integer):
+            info = np.iinfo(spec.dtype)
+            return spaces.Box(info.min, info.max, spec.shape, spec.dtype)
         elif spec.dtype == "bool":
-            low = int(0)
-            high = int(1)
+            return spaces.Box(int(0), int(1), spec.shape, spec.dtype)
         else:
-            raise ValueError(f"Unknown dtype {spec.dtype} for spec {spec}.")
-
-        return spaces.Box(low=low, high=high, shape=spec.shape, dtype=spec.dtype)
-    elif type(spec) is DiscreteArray:
-        return spaces.Discrete(spec.num_values)
+            raise NotImplementedError(f"Unsupported dtype {spec.dtype}")
+    elif isinstance(spec, (list, tuple)):
+        return spaces.Tuple([dm_spec2gym_space(s) for s in spec])
+    elif isinstance(spec, (OrderedDict, dict)):
+        return spaces.Dict(
+            {key: dm_spec2gym_space(value) for key, value in copy.copy(spec).items()}
+        )
     else:
-        raise NotImplementedError(
-            f"Cannot convert dm_spec to gymnasium space, unknown spec: {spec}, please report."
+        raise ValueError(
```

Comment:
TypeError

Reply:
Changed this in the meltingpot branch instead as it updates this function and would lead to merge conflicts if we did it here